src/NuGet.Core/NuGet.Configuration/Settings/AbstractSettingSection.cs

```diff
@@ -0,0 +1,158 @@
+                    }
+                }
+            }
+
+            return this;
+        }
+
+        internal override bool Add(SettingItem setting)
+        {
+            if (setting == null)
+            {
+                throw new ArgumentNullException(nameof(setting));
+            }
+
+            if (!ChildrenSet.ContainsKey(setting) && !setting.IsEmpty())
+            {
+                ChildrenSet.Add(setting, setting);
+
+                return true;
+            }
+
+            return false;
+        }
+
+        internal override void Remove(SettingItem setting)
+        {
+            if (setting == null)
+            {
+                throw new ArgumentNullException(nameof(setting));
+            }
+
+            if (ChildrenSet.TryGetValue(setting, out var currentSetting))
+            {
+                Debug.Assert(!currentSetting.IsAbstract());
+
+                if (currentSetting.Origin.IsMachineWide)
+                {
+                    throw new InvalidOperationException(Resources.CannotUpdateMachineWide);
+                }
+
+                if (ChildrenSet.Remove(currentSetting))
+                {
+                    // Remove it from the appropriate config
+                    if (currentSetting.Parent != null && currentSetting.Parent != this)
+                    {
+                        currentSetting.Parent.Remove(currentSetting);
+                    }
+                }
+
+                if (currentSetting.MergedWith != null)
+                {
+                    // Add that back to the set since, we should leave the machine wide setting intact.
+                    if (!TryRemoveAllMergedWith(currentSetting, out var undeletedItem))
+                    {
+                        ChildrenSet.Add(undeletedItem, undeletedItem);
+                    }
+                }
+            }
+        }
+
+        private bool TryRemoveAllMergedWith(SettingItem currentSetting, out SettingItem undeletedItem)
+        {
+            undeletedItem = null;
+            var mergedSettings = new List<SettingItem>();
+            var mergedWith = currentSetting.MergedWith;
+            while (mergedWith != null)
+            {
+                mergedSettings.Add(mergedWith);
+                mergedWith = mergedWith.MergedWith;
+            }
+
+            foreach (var elementToDelete in mergedSettings)
+            {
+                try
+                {
+                    elementToDelete.Parent.Remove(elementToDelete);
+                }
+                // This means setting was merged with a machine wide settings.
+                catch
+                {
+                    undeletedItem = elementToDelete;
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        internal override SettingBase Clone()
+        {
+            return new AbstractSettingSection(Name, Attributes, Items.Select(s => s.Clone() as SettingItem));
```

Comment:
Wasn't able to follow `Attributes` origin but just make sure that you're doing proper cloning.

Reply:
In our current schema, there are no sections with attributes... This refactor enables the possibility to add attributes in sections... sections inherit from SettingElement which is the one that manages the attributes, there's where this variable comes from.